### PR TITLE
parser: add order field in IndexPartSpecification struct

### LINKED ISSUE #38420: parser: add order field in IndexPartSpecification struct
## Enhancement

> In the past, the parser supported parsing sorting index but did not support it in *.go files, which caused difficulties for related development based on the TiDB parser. And parser is not a separate repository at present, and the way of development after fork becomes very expensive. So we hope to add relevant fields to *.go and mark it at the TiDB level without having to deal with the sorting index.

### PR DESCRIPTION
**Pull request by h3n4l**
### What problem does this PR solve?

close #38420 

Problem Summary:

In the past, the parser supported parsing sorting index but did not support it in *.go files, which caused difficulties for related development based on the TiDB parser. And parser is not a separate repository at present, and the way of development after fork becomes very expensive. So we hope to add relevant fields to *.go and mark it at the TiDB level without having to deal with the sorting index.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: parser/ast/ddl.go
Comment (JmPotato) [inline parser/ast/ddl.go:276]: Forgive my ignorance, do we have an `ASC` in the sorting index?

File: parser/ast/ddl.go
Comment (JmPotato) [inline parser/ast/ddl.go:262]: Better to keep the claim that TiDB is designed to behave consistently with MySQL like before.

Comment (h3n4l) [reply parser/ast/ddl.go:262]: TiDB parser wants to maintain maximum compatibility with MySQL. So it's better to define this struct in Go. So do you think we should change a comment?

Comment (JmPotato) [reply parser/ast/ddl.go:262]: Yeah, I mean it's better to retain the original comment "// Order is parsed but just ignored as MySQL did."

File: parser/ast/ddl.go
Comment (JmPotato) [reply parser/ast/ddl.go:276]: Got it, thx for explanation.

File: parser/ast/ddl.go
Comment (h3n4l) [reply parser/ast/ddl.go:262]: Hey @JmPotato , I think this comment had overdue.
MySQL parser parses order index in v5.x, but it will be ignored because MySQL doesn't support it in v5.x. But since MySQL 8.0, MySQL will not ignore it because it had already been supported.
I prefer to use `// Order is parsed but should be ignored because TiDB doesn't support it.`, how about you?

Comment (JmPotato) [reply parser/ast/ddl.go:262]: Since TiDB claims to be compatible with MySQL v5.7 according to https://docs.pingcap.com/tidb/dev/mysql-compatibility, so I think it's OK to level the original comment. But actually, my core idea is that the parser is some kind of a public dependency, and we can probably avoid introducing things like "TiDB" into it as much as possible. Using "Order is parsed but should be ignored because MySQL v5.7 doesn't support it" sounds more natural to me. WDYT?

Comment (h3n4l) [reply parser/ast/ddl.go:262]: LGTM! I will modify it.

Comment (h3n4l) [reply parser/ast/ddl.go]: Done. PTAL @JmPotato

**Review by JmPotato — APPROVED** [review]

**Comment by JmPotato** [comment]
/cc @xhebox 

Could you please help review this PR?

**Review by xhebox — COMMENTED** [review]
LGTM, maybe it is worth notifying tools team.

BTW, do we need any thing else since this is noticeable by users? /cc @bb7133

**Comment by h3n4l** [comment]
> LGTM, maybe it is worth notifying tools team.
> 
> BTW, do we need any thing else since this is noticeable by users? /cc @bb7133

If TiDB team wants to make the parser more compatible, I can try to do something like this PR.

**Comment by h3n4l** [comment]
@JmPotato Could you please help me merge this PR?

**Comment by JmPotato** [comment]
> @JmPotato Could you please help me merge this PR?

Merging requires 1 more maintainer's approval, I think we need to wait for @bb7133 to review.

**Review by xhebox — APPROVED** [review]
I think it is OK.

/merge

**Pull request merged**
